### ISSUE TITLE
What are the ways of treating missing values in tf-ranking?

### ISSUE THREAD

**Issue by primetang** [tensorflow/ranking#58]

**Comment by bendersky** [comment]
Gefu, can you be more specific about your question? Right now, if your input format is libsvm you can only specify the features that are present and ignore the missing features.

**Comment by primetang** [comment]
[In the example](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_libsvm.py), the input format is libsvm, and the missing features is treated as zero, it will be treated like other features with a value of 0., but I don't think it is suitable. When a feature is missing, it indicates that we don't know its value, but not 0. I think the ways of treatng missing values in XGboost is more reasonable.

[By @tqchen](https://github.com/dmlc/xgboost/issues/21): 

xgboost naturally accepts sparse feature format, you can directly feed data in as sparse matrix, and only contains non-missing value.

i.e. features that are not presented in the sparse feature matrix are treated as 'missing'. XGBoost will handle it internally and you do not need to do anything on it.

Internally, XGBoost will automatically learn what is the best direction to go when a value is missing. Equivalently, this can be viewed as automatically "learn" what is the best imputation value for missing values based on reduction on training loss.

**Comment by primetang** [comment]
> Gefu, can you be more specific about your question? Right now, if your input format is libsvm you can only specify the features that are present and ignore the missing features.

[In the example](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_libsvm.py), the input format is libsvm, and the missing features is treated as zero, it will be treated like other features with a value of 0., but I don't think it is suitable. When a feature is missing, it indicates that we don't know its value, but not 0. I think the ways of treatng missing values in XGboost is more reasonable.

[By @tqchen](https://github.com/dmlc/xgboost/issues/21):

xgboost naturally accepts sparse feature format, you can directly feed data in as sparse matrix, and only contains non-missing value.

i.e. features that are not presented in the sparse feature matrix are treated as 'missing'. XGBoost will handle it internally and you do not need to do anything on it.

Internally, XGBoost will automatically learn what is the best direction to go when a value is missing. Equivalently, this can be viewed as automatically "learn" what is the best imputation value for missing values based on reduction on training loss.

**Comment by bendersky** [comment]
Thanks for the explanation. At this point we do not provide any specific imputation techniques as a part of the TF-Ranking core library.

**Comment by xuanhuiwang** [comment]
To add to what bendersky@ has said, in TensorFlow, each Tensor has a type and the value can not be None. Thus a default value like 0 has to be put there for a missing value.

One way to handle missing feature is to have an additional feature "is_feature_a_missing" and put it as 1 if feature_a is missing and 0 otherwise. In feature_a, 0 could mean value 0 or missing feature. The corresponding is_feature_a_missing help disambiguate them.

**Comment by primetang** [comment]
> To add to what bendersky@ has said, in TensorFlow, each Tensor has a type and the value can not be None. Thus a default value like 0 has to be put there for a missing value.
> 
> One way to handle missing feature is to have an additional feature "is_feature_a_missing" and put it as 1 if feature_a is missing and 0 otherwise. In feature_a, 0 could mean value 0 or missing feature. The corresponding is_feature_a_missing help disambiguate them.

Understand! thank you.

**Comment by bendersky** [comment]
Closing this for now, please feel free to re-open if you have any additional questions.
